src/modules/roc_netio/target_uv/roc_netio/transceiver.cpp

```diff
@@ -51,92 +80,50 @@ Transceiver::Transceiver(packet::PacketPool& packet_pool,
     task_sem_.data = this;
     task_sem_initialized_ = true;
 
-    valid_ = true;
+    started_ = Thread::start();
 }
 
 Transceiver::~Transceiver() {
-    if (joinable()) {
-        roc_panic("transceiver: thread is not joined before calling destructor");
+    if (stop_sem_initialized_) {
+        if (int err = uv_async_send(&stop_sem_)) {
+            roc_panic("transceiver: uv_async_send(): [%s] %s", uv_err_name(err),
+                      uv_strerror(err));
+        }
+    } else if (task_sem_initialized_) {
+        close_sems_();
     }
 
-    if (num_ports_ != 0) {
-        remove_all_ports_();
+    if (started_) {
+        Thread::join();
+    } else {
```

Comment:
And here we should replace `else {` with `else if (loop_initialized_) {`. Or, alternatively, we can rewrite this piece as:

```
if (loop_initialized_) {
    if (started_) {
        Thread::join();
    } else {
        // If the thread was never started we should manually run the loop to
        // wait all opened handles to be closed. Otherwise, uv_loop_close()
        // will fail with EBUSY.
        Transceiver::run(); // non-virtual call from dtor
    }

    
    if (int err = uv_loop_close(&loop_)) {
        roc_panic("transceiver: uv_loop_close(): [%s] %s", uv_err_name(err),
                  uv_strerror(err));
    }
}
```

Reply:
What if loop was not initialized but stop_sem and task_sem were initialized?

Reply:
It's not possible. To initialize any handle, we need an initialized loop.

Reply:
Done